### PR TITLE
Automatically save replays for global warnings and autolocks

### PR DESCRIPTION
**Pull request by asgdf**
There is of course the problem that the savereplay command requires a user, which does not mesh well with the concept of an autolock. Current convention seems to be to just use the user who is getting locked, which works, but means they'll get the replay sent to them with the popup and all. I don't really like that but I'm not sure what else to do.

### CODE REVIEW

File: server/punishments.ts [hunk 1]
```diff
@@ -679,6 +679,7 @@ export const Punishments = new class {
 		const ipStr = typeof user !== 'string' ? ` [${(user as User).latestIp}]` : '';
 		const roomid = typeof room !== 'string' ? (room as Room).roomid : room;
 		Rooms.global.modlog(`(${roomid}) AUTO${namelock ? `NAME` : ''}LOCK: [${userid}]${ipStr}: ${reason}`);
+		Chat.parse('/savereplay', room as Room, user as User, (user as User).connections[0]);
```
Comment (whalemer) [inline server/punishments.ts]: this cast isn't always right, see: https://github.com/Zarel/Pokemon-Showdown/blob/c990e015dbff2cece541833e77cb4a903a61cd2a/server/punishments.ts#L1452

Comment (asgdf) [reply server/punishments.ts]: Oh that's fair.

File: server/punishments.ts [hunk 2]
```diff
@@ -679,6 +679,7 @@ export const Punishments = new class {
 		const ipStr = typeof user !== 'string' ? ` [${(user as User).latestIp}]` : '';
 		const roomid = typeof room !== 'string' ? (room as Room).roomid : room;
 		Rooms.global.modlog(`(${roomid}) AUTO${namelock ? `NAME` : ''}LOCK: [${userid}]${ipStr}: ${reason}`);
+		Chat.parse('/savereplay', Rooms.get(room), user as User, (user as User).connections[0]);
```
Comment (whalemer) [inline server/punishments.ts]: This still isn't right, you can trigger this with a userid (ie if you nameblacklist someone with 2 RBs). You also need to make sure that `connections[0]` is actually defined (ie, punishing an offline user)

Comment (asgdf) [reply server/punishments.ts]: I didn't even check whether the room was a battle at all. What was I thinking? I guess I'm dumb.

File: server/punishments.ts [hunk 3]
```diff
@@ -679,6 +679,12 @@ export const Punishments = new class {
 		const ipStr = typeof user !== 'string' ? ` [${(user as User).latestIp}]` : '';
 		const roomid = typeof room !== 'string' ? (room as Room).roomid : room;
 		Rooms.global.modlog(`(${roomid}) AUTO${namelock ? `NAME` : ''}LOCK: [${userid}]${ipStr}: ${reason}`);
+
+		const roomObject = Rooms.get(room);
+		const userObject = Users.get(user);
+		if (roomObject.battle && userObject && userObject.connections[0]) {
```
Comment (whalemer) [inline server/punishments.ts]: Not checking if `roomObject` exists (the return type for getRoom is wrong) makes me nervous, but I think it's always going to be fine in practice.

Comment (Zarel) [reply server/punishments.ts]: Well, it's going to need a TS override now that we merged your PR. :p

File: server/punishments.ts [hunk 4]
```diff
@@ -679,6 +679,12 @@ export const Punishments = new class {
 		const ipStr = typeof user !== 'string' ? ` [${(user as User).latestIp}]` : '';
 		const roomid = typeof room !== 'string' ? (room as Room).roomid : room;
 		Rooms.global.modlog(`(${roomid}) AUTO${namelock ? `NAME` : ''}LOCK: [${userid}]${ipStr}: ${reason}`);
+
+		const roomObject = Rooms.get(room);
+		const userObject = Users.get(user);
+		if (roomObject && roomObject.battle && userObject && userObject.connections[0]) {
```
Comment (Zarel) [inline server/punishments.ts]:
```js
		if (roomObject?.battle && userObject?.connections[0]) {
```

wtb optional chaining ;_;

**Comment by Zarel** [comment]
> Current convention seems to be to just use the user who is getting locked, which works, but means they'll get the replay sent to them with the popup and all.

Wait, what???? I thought current convention was to use the mod who did the locking........

File: server/punishments.ts [hunk 5]
```diff
@@ -679,6 +679,12 @@ export const Punishments = new class {
...
+			Chat.parse('/savereplay', roomObject, userObject, userObject.connections[0]);
```
Comment (Zarel) [inline server/punishments.ts]: Why not `/savereplay forPunishment` like locks?

Comment (Zarel) [reply server/punishments.ts]: Without it, hidden room replays are inaccessible to mods.

**Comment by asgdf** [comment]
Okay, removed the auth check in /savereplay forpunishment, and made autolocks use them.

**Pull request merged**